src/SMT.hs

```diff
@@ -0,0 +1,446 @@
+import Print (prettyEnv)
+import Type (defaultStore, metaType)
+
+import System.Process (readProcessWithExitCode)
+import System.Exit (ExitCode(..))
+
+--- Data ---
+
+data Solver = Z3 | CVC4
+  deriving Eq
+
+instance Show Solver where
+  show Z3 = "z3"
+  show CVC4 = "cvc4"
+
+data SMTConfig = SMTConfig
+  { _solver :: Solver
+  , _timeout :: Integer
+  , _debug :: Bool
+  }
+
+type SMT2 = String
+
+-- | Some expressions should be constructed over the poststate, and some over the prestate...
+data When = Pre | Post
+
+instance Show When where
+  show Pre = "pre"
+  show Post = "post"
+
+-- | An SMTExp is a structured representation of an SMT Expression
+--   The _storage, _calldata, and _environment fields hold variable declarations
+--   The assertions field holds the various constraints that should be satisfied
+data SMTExp = SMTExp
+  { _storage :: [SMT2]
+  , _calldata :: [SMT2]
+  , _environment :: [SMT2]
+  , _assertions :: [SMT2]
+  }
+
+instance Show SMTExp where
+  show e = intercalate "\n" ["", storage, "", calldata, "", environment, "", assertions, ""]
+    where
+      storage = ";STORAGE:\n" <> intercalate "\n" (nubOrd $ _storage e)
+      calldata = ";CALLDATA:\n" <> intercalate "\n" (nubOrd $ _calldata e)
+      environment = ";ENVIRONMENT:\n" <> intercalate "\n" (nubOrd $ _environment e)
+      assertions = ";ASSERTIONS:\n" <> intercalate "\n" (nubOrd $ _assertions e)
+
+-- A Query is a structured representation of an SMT query for an individual
+-- expression, along with the metadata needed to pretty print the result
+data Query
+  = Postcondition Claim (Exp Bool) SMTExp
+  | Inv Claim Invariant SMTExp
+  deriving (Show)
+
+data SMTResult
+  = Sat
+  | Unsat --Model
+  | Unknown
+  | Error Int String
+  deriving (Show)
+
+data Model = Model
+  { _mstorage :: Map Id (MType, MType)
+  , _mcalldata :: Map Id MType
+  , _menvironment :: Map Id MType
+  }
+  deriving (Show)
+
+--- External Interface ---
+
+-- | For each postcondition in the claim we construct a query that:
+--    - Asserts that the preconditions hold
+--    - Asserts that storage has been updated according to the rewrites in the behaviour
+--    - Asserts that the postcondition cannot be reached
+--   If this query is unsatisfiable, then there exists no case where the postcondition can be violated.
+mkPostconditionQueries :: Claim -> [Query]
+mkPostconditionQueries (B behv@(Behaviour _ _ _ (Interface _ decls) preconds postconds stateUpdates _)) = mkQuery <$> postconds
+  where
+    -- declare vars
+    storage = concatMap (declareStorageLocation . getLoc) stateUpdates
+    args = declareArg <$> decls
+    envs = declareEthEnv <$> ethEnvFromBehaviour behv
+
+    -- constraints
+    pres = mkAssert Pre <$> preconds
+    updates = encodeUpdate <$> stateUpdates
+
+    mksmt e = SMTExp
+      { _storage = storage
+      , _calldata = args
+      , _environment = envs
+      , _assertions = pres <> updates <> [mkAssert Pre . Neg $ e]
```

Comment:
Not super important but since lists in Haskell are linked it's best if the first argument to `<>` is as short as possible.

```suggestion
      , _assertions = [mkAssert Pre . Neg $ e] <> pres <> updates
```